### PR TITLE
Fix bug 48516

### PR DESCRIPTION
**Pull request by petertiedemann**
My attempt to fix: https://bugzilla.xamarin.com/show_bug.cgi?id=48516

### CODE REVIEW

File: mcs/class/System.IO.Compression/Test/System.IO.Compression/ZipTest.cs [hunk 1]
```diff
@@ -507,6 +507,11 @@ class MyFakeStream : FileStream
 			/// Simulate "CanSeek" is false, which is the case when you are retreiving data from web.
 			/// </summary>
 			public override bool CanSeek => false;
+
+			public override long Position {
```
Comment (petertiedemann) [inline mcs/class/System.IO.Compression/Test/System.IO.Compression/ZipTest.cs]: Non-seekable streams typically do not support even accessing the position property.

File: mcs/class/System.IO.Compression/Test/System.IO.Compression/ZipTest.cs [hunk 2]
```diff
@@ -517,5 +522,16 @@ public void ZipReadNonSeekableStream()
...
 			}
 		}
+
+		[Test]
+		public void ZipWriteNonSeekableStream() {
+			var stream = new MyFakeStream( "test.nupkg", FileMode.Open );
+			using ( var archive = new ZipArchive( stream, ZipArchiveMode.Create ) ) {
+				var entry = archive.CreateEntry( "foo" );
+				using ( var es = entry.Open() ) {
+					es.Write( new byte[] { 4, 2 }, 0, 2 );
```
Comment (petertiedemann) [inline mcs/class/System.IO.Compression/Test/System.IO.Compression/ZipTest.cs]: Test showing the problem. In Create mode, ZipArchive would previously attempt to set both Position and Length on the stream.

File: mcs/class/System.IO.Compression/ZipArchive.cs [hunk 3]
```diff
@@ -227,13 +227,20 @@ public ZipArchiveEntry GetEntry (string entryName)
 
 		private void Save()
 		{
-			using (var newZip = new MemoryStream()) {
-				zipFile.SaveTo(newZip, CompressionType.Deflate, entryNameEncoding ?? Encoding.UTF8);
+			if (mode == ZipArchiveMode.Create)
+			{
+				zipFile.SaveTo(stream, CompressionType.Deflate, entryNameEncoding ?? Encoding.UTF8);
```
Comment (petertiedemann) [inline mcs/class/System.IO.Compression/ZipArchive.cs]: If i understand correctly, then the reason for saving to a MemoryStream first, is that the underlying zipFile might try to read from the stream in update mode. But in Create mode we are not reading from the output stream.

**Comment by migueldeicaza** [comment]
While it looks fine to me, I would like Joao to review this.

**Comment by tritao** [comment]
Looks good, thanks for sending the PR.

**Pull request merged**